### ISSUE TITLE
fix printing of numerical fields

### ISSUE THREAD

**Issue by GitMensch** [OlegKunitsyn/gnucobol-debug#30]
This is something that especially is important for non-display fields and was initially brought up in #4. The information about those fields and their attributes is already in the generated code, some hints on this can be found in https://github.com/OlegKunitsyn/gnucobol-debug/issues/4#issuecomment-627829938.

**Comment by brunopacheco1** [comment]
Hi @OlegKunitsyn Could checkout the latest changes on fields branch? Numeric field are being parsed now.

@GitMensch Could you provide a COBOL sample presenting all possible data types?

**Comment by OlegKunitsyn** [comment]
Merged into master. Thank you!

**Comment by GitMensch** [comment]
@OlegKunitsyn Thank you for your work, here's your requested test sample in exchange (_quite_ complete):

```cobol
       IDENTIFICATION   DIVISION.
       PROGRAM-ID.      types.
       DATA             DIVISION.
       WORKING-STORAGE  SECTION.
       01 numeric-data.
          05 disp       usage display        pic s99v999   value -12.34.
          05 disp-u     usage display        pic  99v999   value  12.34.
          05 dispp      usage display     pic spppp9999 value -.0000123.
          05 dispp-u    usage display     pic  pppp9999 value  .0000123.
          05 disppp     usage display     pic s9999pppp value -12340000.
          05 disppp-u   usage display     pic  9999pppp value  12340000.
          05 bin        usage binary         pic s99v999   value -12.34.
          05 bin-u      usage binary         pic  99v999   value  12.34.
          05 cmp3       usage packed-decimal pic s99v999   value -12.34.
          05 cmp3-u     usage packed-decimal pic  99v999   value  12.34.
          05 cmp5       usage comp-5         pic s99v999   value -12.34.
          05 cmp5-u     usage comp-5         pic  99v999   value  12.34.
          05 cmp6       usage comp-6         pic  99v999   value  12.34.
          05 cmpx       usage comp-x         pic s99v999   value -12.34.
          05 cmpx-u     usage comp-x         pic  99v999   value  12.34.
          05 cmpn       usage comp-n         pic s99v999   value -12.34.
          05 cmpn-u     usage comp-n         pic  99v999   value  12.34.
          05 chr        usage binary-char    signed   value -128.
          05 chr-u      usage binary-char    unsigned value 254.
          05 shrt       usage binary-short   signed   value -32768.
          05 shrt-u     usage binary-short   unsigned value 65535.
          05 long       usage binary-long    signed   value -2147483648.
          05 long-u     usage binary-long    unsigned value  4294967295.
          05 dble       usage binary-double  signed   value -4294967295.
          05 dble-u     usage binary-double  unsigned value  8294967295.
      *   05 cmp0       usage comp-0.  TODO
       01 floating-data.
          05 dbl        usage float-long      value -3.40282e+038.
          05 flt        usage float-short     value 3.40282e+038.
      *   05 b32        usage FLOAT-BINARY-32  signed.      TODO
      *   05 b64        usage FLOAT-BINARY-64  signed.      TODO
      *   05 b128       usage FLOAT-BINARY-128 signed.      TODO
      *   05 b32-u      usage FLOAT-BINARY-32  unsigned.    TODO
      *   05 b64-u      usage FLOAT-BINARY-64  unsigned.    TODO
      *   05 b128-u     usage FLOAT-BINARY-128 unsigned.    TODO
      *
      *  --> missing:   floating point edited pictures,
      *                 those are left to the reader as excercise :-)
       01 special-data.
          05 r2d2       usage bit pic 111 value b'110'.
          05 point      usage pointer.
          05 ppoint     usage program-pointer.
          05 idx        usage index.
          05 hnd        usage handle.
       01 alphanumeric-data.
          05 alpnum     pic x(36) value "some numb3rs 4 n00bs l1k3 m3".
          05 alpha      pic a(36) value "that's some text".
          05 edit-num1  pic --9.999.
          05 edit-num2  pic ++9.999.
          05 edit-num3  pic zz9.999.
          05 edit-num4  pic -zz9.999   blank when zero.
       01 national-data.
          05 nat        pic n(36) value "data shown here will change.".
          05 nat-num    pic 9(12)v9(3) usage national.
          05 net-num1   pic --9.999    usage national.
          05 net-num2   pic ++9.999    usage national.
          05 net-num3   pic zz9.999    usage national.
          05 net-num4   pic -zz9.999   usage national blank when zero.
       PROCEDURE DIVISION.
           set  idx    to 1
           set  point  to address of ppoint
           set  ppoint to entry "types"
           move cmp5   to edit-num1, edit-num2, edit-num3, edit-num4
           move cmp5   to net-num1,  net-num2,  net-num3,  net-num4

           display disp
           display disp-u
           display dispp
           display dispp-u
           display disppp
           display disppp-u
           display bin
           display bin-u
           display cmp3
           display cmp3-u
           display cmp5
           display cmp5-u
           display cmp6
           display cmpx
           display cmpx-u
           display cmpn
           display cmpn-u
           display chr
           display chr-u
           display shrt
           display shrt-u
           display long
           display long-u
           display dble
           display dble-u
           display dbl
           display flt
           display r2d2
           display point
           display ppoint
           display idx
           display hnd
           display alpnum
           display alpha
           display edit-num1
           display edit-num2
           display edit-num3
           display edit-num4
           display nat
           display net-num1
           display net-num2
           display net-num3
           display net-num4

           display numeric-data
           display floating-data
           display special-data
           display alphanumeric-data
           display national-data

           GOBACK.
```

**Comment by brunopacheco1** [comment]
@GitMensch Thanks a lot for this sample, there's quite a lot of work to do on that.

**Comment by brunopacheco1** [comment]
So, basically, all alphanumeric and national data-typed fields are working properly. All floating and special data-typed are failing and 25% of numeric data-typed fields are working.

I pushed some minor fixes thanks to this sample and I'll keep working on them, but I believe all basic and most common data types are covered.

**Comment by GitMensch** [comment]
For the national types (where GnuCOBOL warns that the implementation is unfinished) the content is expected to be in UTF-16, I don't know how good/bad GDB is in display/accept of those values. As a real test you may move some "real utf-16 data" in there by changing the definition to
```cobol
          05 nat        pic n(36)
             value nx"0054006500730074002000E420AC".
```

Thank you for working on the numerics. For some details you may want to check GnuCOBOL's libcob/numeric.c which provides internal routines to pack/unpack.

**Comment by GitMensch** [comment]
I've updated to the current version and gave it a quick check - in general it looks *much better* now.
But there are some strange things going on with the attached test code (if `-fdump` is not used, if it is used we get a compile error, so there's something to fix in the GnuCOBOL part, too).

![variable view](https://user-images.githubusercontent.com/6699539/82559400-da0a3a80-9b6f-11ea-8271-c558e79c917d.png)

The most strange thing: IDX1 is shown as group containing only itself, similar for the actual group in `HEADING-LINE`. I _guess_ the missing `DIGS` will show with the dump code enabled.

[test.cob](https://github.com/OlegKunitsyn/gnucobol-debug/files/4662431/test.cob.txt)

**Comment by brunopacheco1** [comment]
Yes, when I first saw that I thought it was a bug, but actually GC adds a data storage variable (cob_u8_t) and a field  (cob_field) with the same name.

My first idea was to present the field value within the data storage, replacing the type by the value, then I thought to ignore the field and not present the data storage value at all, but at the end presenting it as a child node looked less polluted and quite handy to confirm the children's value when debugging the extension.

Definitely adding `-dump=ALL` it will present most of the variables.

**Comment by GitMensch** [comment]
> Yes, when I first saw that I thought it was a bug, but actually GC adds a data storage variable (cob_u8_t) and a field (cob_field) with the same name.

not exactly: it will do so _for_ the same name. As soon as you can parse the field I'd completely remove the data storage variable. Why do you want to keep it?

**Comment by OlegKunitsyn** [comment]
Set breakpoint at subsample.cbl line15. Run sample.cbl. WS-GROUP:Group in content 'Invalid variable attributes' displayed. Is it expected?

**Comment by OlegKunitsyn** [comment]
Set breakpoints in func.cbl at 15, 27, 39. Run and expand variables in the panel:
- Cannot read property...
- No symbol...

**Comment by brunopacheco1** [comment]
@OlegKunitsyn Both bugs were fixed.

For the first the variable panel is not expecting real null values when that happens. To fix, the extension is returning a string `"null"`.

For the second, my assumption was the GC would generate different `*.c.l.h` for all functions but actually when the functions are in the same file it generates that files with this pattern `prog.c.l[0-9]*.h`.

**Comment by brunopacheco1** [comment]
@GitMensch I put the panel to display the data storage fields as a "Value" child as it used to be before and the data storage cob_u8_t is ignored if there is the equivalent of cob_field.

Some minor fixes were done on parsing and displaying GC data types, but no further new implementation. I am struggling to understand the GC code and I have a feeling that reimplementing the cob_display function in this extension is not the best practice.

I would say the ideal solution would be reusing some GC function to get the display value of the given cob_field or something similar to `printf`, just giving the format and the cob_field.

These are some samples of binary values coming from GDB:

![image](https://user-images.githubusercontent.com/4491850/82742324-edc6c400-9d5c-11ea-85f9-5aab46a369dc.png)

**Comment by OlegKunitsyn** [comment]
@brunopacheco1 I guess we may release the latest changes. Can you merge the branch with the master please? There is 10 conflicts.

**Comment by GitMensch** [comment]
> I put the panel to display the data storage fields as a "Value" child as it used to be before and the data storage cob_u8_t is ignored if there is the equivalent of cob_field.
> 
> Some minor fixes were done on parsing and displaying GC data types,

That sounds very good!

> I am struggling to understand the GC code and I have a feeling that reimplementing the cob_display function in this extension is not the best practice.
> 
> I would say the ideal solution would be reusing some GC function to get the display value of the given cob_field or something similar to `printf`, just giving the format and the cob_field.

What do you think of the functions [outlined before](https://github.com/OlegKunitsyn/gnucobol-debug/issues/4#issuecomment-626041715)?

Can you `call (int)cob_get_num_params` on the debug console (should return the number of the `USING` parameters passed, so likely 0 for any "main" program) and in the plugin? If yes - what about `call (char *)cob_getenv("PATH")`? If this works fine something like the functions outlined before may be useful.

**Comment by brunopacheco1** [comment]
@OlegKunitsyn the merge is done.

**Comment by brunopacheco1** [comment]
@GitMensch I have been working on a similar solution but using `display_common (cob_field *f, FILE *fp)`, which is a function present in both versions.

So, the solution is quite hacky as the extension is printing on stdout a fake MI event using a GDB call command (e.g. `555^done,value=|-3.40282e+038`, and the extension is capturing and handling it as any other MI event.

As you can see in the print-screen below, it looks quite good. 

![image](https://user-images.githubusercontent.com/4491850/82846859-37ddb000-9eeb-11ea-8a0a-5ba21056ebdd.png)

In most cases this solution is working without issues, but depending on the COBOL sample and depending on where the breakpoint is, the functions `display_common` throws a segmentation fault, and everything crashes. There is a workaround by forcing GDB to restore the previous state, but then all previous commands are re-executed and then the extension asks again for the failing variable and then the crash happens again.

The loop is not a problem if we put a delay of a couple of secs and if the user keeps stepping over until a step where the segmentation fault doesn't happen. The user can see the variable panel has gone to a loop, but the debugger is still usable.

**Comment by brunopacheco1** [comment]
I believe the best solution is somewhere in between. Let me know what you think guys.

If the GC is version 3.1+ we call the function `cob_print_field`, otherwise if it is numeric and alphanumeric we keep using the existing method, and if it is binary we call `display_common`.

**Comment by brunopacheco1** [comment]
BTW, we are still missing TABLEs and I have no idea how to tackle it.

**Comment by OlegKunitsyn** [comment]
Thank you very much! IMO so far so good. I would close the thread, but TABLE in new FR and release. Deal?

**Comment by GitMensch** [comment]
Depends, if there's a SIGSEGV anywhere is say no close, no release. Otherwise a release is useful.
If only OCCURS is an issue I'd say yes to closing, but again a no for missing a bunch of USAGEs, which seems the case.
While this may be tracked in a different issue: what is the status of altering the variables?

**Comment by GitMensch** [comment]
> I have been working on a similar solution but using `display_common (cob_field *f, FILE *fp)`, which is a function present in both versions.

I like this a lot, I wasn't sure about the question of this being accessible as it is a static function, local to libcob/termio.c.
We could either externalize this, with some security checks added or add the new (not yet available) `cob_print_field()`.

> So, the solution is quite hacky as the extension is printing on stdout a fake MI event using a GDB call command (e.g. `555^done,value=|-3.40282e+038`, and the extension is capturing and handling it as any other MI event.

Sounds like this should also work in the future, "hack like this" is good.

> In most cases this solution is working without issues, but depending on the COBOL sample and depending on where the breakpoint is, the functions `display_common` throws a segmentation fault, and everything crashes. There is a workaround by forcing GDB to restore the previous state, but then all previous commands are re-executed and then the extension asks again for the failing variable and then the crash happens again.

From a quick glance: it should always work as long as you pass an actual field with field->data set (both not NULL). An externalized version of this could check for both and "display" `NULL field` or `NULL data` if those aren't valid, but as you want to do it in 2.2, too I suggest you add those checks.

If you can then still reproduce the crash please provide a sample and I'd try to debug the issue in libcob.

**Comment by brunopacheco1** [comment]
Sorry guys, I didn't have time to check the last comments, but tomorrow I will address them.

**Comment by brunopacheco1** [comment]
Well let's start.

I've been working only on displaying data so far. As explained before, the idea was to use something coming from libcob to avoid rewriting functions. And for that, I am using `display_common (cob_field *f, FILE *fp)`, printing a fake MI event, which is seamlessly understood by the extension.

I thought initially it would be a safer choice keeping the old behavior but actually adding nullability check before calling display_common as suggest by @GitMensch was enough to ensure that no crash is going to happen.

I committed the changes to the branch fields.

**Comment by brunopacheco1** [comment]
So, to sum up all data types present in the initial sample are being displayed on the variable panel as they would be presented on the console, and I believe any other data type not included here would be presented in the same way as long as both projects (extension and GC) keep using the same function `display_common` .

I will start working on setting variable values, and for me doesn't change opening a new issue or keep using this one. The only bennefit of moving to a new one is a clean log.

**Comment by GitMensch** [comment]
I've created a new issue for the setting #40 and suggest to adjust this issues's title and close it as soon as the work is part of master.

> So, to sum up all data types present in the initial sample are being displayed on the variable panel as they would be presented on the console, and I believe any other data type not included here would be presented in the same way as long as both projects (extension and GC) keep using the same function `display_common` .

Thank you, well done!

Note: `display_common` is not part of the public API and therefore _possibly_ be changed at any time. Please create a [GnuCOBOL FR](https://sourceforge.net/p/open-cobol/feature-requests/new/) to add an external version (likely would be named `cob_field_display` or something like that - as it is a wish tracker you can be very specific what you'd like to have [it will include the null checks, for example and output something special in this case] :-) - you may also need something for the setting of fields, if you have an idea please drop it there, too).

**Comment by brunopacheco1** [comment]
@GitMensch I =requested the features as suggested. Thanks for all support you've been giving.

**Comment by OlegKunitsyn** [comment]
Please compare the values with
https://raw.githubusercontent.com/OlegKunitsyn/gnucobol-debug/master/screenshot.png
I see nulls since Commit 291a30fe.

**Comment by brunopacheco1** [comment]
Hi,  @OlegKunitsyn Could you share the full log?

Have you defined expressions on the WATCH panel? I'm asking because there are few possibilities for a cob field evaluation to be called, either expanding the field on VARIABLES panel, having a WATCH expression, or hovering over a variable.

So, my guess is the libcob wasn't loaded properly or an error was thrown during runtime before this cob field evaluation.

**Comment by brunopacheco1** [comment]
I debugged petstore sample and all the others, and it seems all of them are running, the displaying variables feature is not working correctly.

If I am running the extension in debug mode and I put a breakpoint before the command `stack-list-variables`, variables are fetched but not displayed. If I put a breakpoint after, not even the fetching happens.

I believe we should investigate the changes around `getStackVariables`, instead, as the bug is consistently happening before cob field evaluation, at least for me.

**Comment by OlegKunitsyn** [comment]
I use GC 2.2, the latest stable, available in the distribution since the first commit in this project. 

[hello.txt](https://github.com/OlegKunitsyn/gnucobol-debug/files/4709425/hello.txt)
[petstore.txt](https://github.com/OlegKunitsyn/gnucobol-debug/files/4709426/petstore.txt)

![Screenshot from 2020-06-01 06-47-06](https://user-images.githubusercontent.com/2870028/83376838-f90c9600-a3d3-11ea-9fc2-404bb83b5f2e.png)

**Comment by OlegKunitsyn** [comment]
GC 3.1-dev nightly fresh:
![Screenshot from 2020-06-01 07-32-15](https://user-images.githubusercontent.com/2870028/83379127-13963d80-a3db-11ea-9ed1-3baf27026995.png)

**Comment by brunopacheco1** [comment]
Hi @OlegKunitsyn I've been debugging on both GC 2.2 and 3.1-dev, locally and on Docker, so we have two issues, first, it looks like display_common function is not accessible on GC 2.2. The second problem is on Docker, for some reason calling `fflush` to clear the buffer, forcing stdout to trigger the changes is not working as it should.

For the first one the fix is simple, I am using cob_display instead, I don't know why I didn't stick to that instead of display_common, but it is done.

For the second problem I am not sure why that is happening, but to unblock the master I will stop using `fflush` and leave the cob_display function to do flushing job. The only drawback is I cannot append at the end of the cob_display result and `"`, which will break the MI event parsing. So, the workaround is to have a new parser to understand a customized MI event for cob field evaluation.

**Comment by OlegKunitsyn** [comment]
Yes, it's getting complicated. Thank you for not despairing! :) You know, I would support single environment, but 2.2. already in the wild. And Docker image is only way to deliver new features to the platforms bypassing an indefinite process of "normal" distribution. Good news is we can build, bundle deliver any Docker image. One day we see COBOL OOP there. I need it badly, @GitMensch. :)
Since `fflush` so important I'll backtrace Docker stuff.

**Comment by GitMensch** [comment]
> first, it looks like display_common function is not accessible on GC 2.2

The signature has not changed, in both cases it is a non-exported static local function. Which was the reason I've asked to create the [GnuCOBOL FR Printing cob_field values using libcob](https://sourceforge.net/p/open-cobol/feature-requests/383/). It _may_ be accessible but this is not something one can rely to. If it it works on GC3.1dev dockerized the docker environment should use this one (which I think it does).

> I am using cob_display instead, I don't know why I didn't stick to it instead of display_common, but it is done.

This has a little overhead and the only guaranteed way to not have the output somewhere externally redirected is to call it as `cob_display (1, 0, 1, field_ptr)` - @brunopacheco1 is this done now?

> So, the workaround is to have a add to the MI parser a customized MI event marker for cob field evaluation.

Sounds reasonable in any case.

**Comment by brunopacheco1** [comment]
@GitMensch Nope, I put `cob_display(0,1,1,field_pointer)`, as on GC 2.2 is defined in the following way `cob_display (const int to_stderr, const int newline, const int varcnt, ...)`. I thought we should print it on STDOUT, not on STDERR. And it has to be a new line, because fflush is not working properly, I still have to investigate it more.

**Comment by brunopacheco1** [comment]
As we still have a issue related to that #35 , this is not finished yet.

**Comment by GitMensch** [comment]
1 is STDERR in GC3+, too. Every other option may be placed "anywhere", depending on the actual setup (environment variables, runtime configuration, previous calls to C ABI), therefore I suggest to always use `1` if you use `cob_display` (much less error prone); no idea if/how GDB can read events from there...

**Comment by brunopacheco1** [comment]
I believe it's safer keeping print everything on STDOUT for a while and tackle this topic on #26 .

But after these errors I agree that the extension should redirect different outputs to different panels, one single panel for everything is confusing and becoming more and more polluted. each new feature.

**Comment by GitMensch** [comment]
As noted: 0 != STDOUT, I may wasn't clear on this. 0 = STDOUT if there was already an extended screenio not configured otherwise (by runtime settings via environment, runtime configuration or by a call via C ABI).

You can easily test this by:

* setting `COB_REDIRECT_DISPLAY=1` before
* add `DISPLAY "test goes on" at 0101` at the program start

both will break the current implementation, and especially the second _should_ not be the case. Please adjust before the next release.

**Comment by brunopacheco1** [comment]
Well, the more I try to figure out how to attach to a running process (#3 ), the more I believe the way we are getting cob_field values has to change, we should not rely on another source besides GDB.

The current variable evaluation is relying on the Debug Console, and at the first sight it looks handful and give us the impression of a silver bullet code because we are supporting tons of different usage cases, but it causes more issues than helping.

Unfortunately, this solution is tightly coupled with the environment (terminal), even if we find a way to get the inferior program terminal and read the outputs generated there, this is not going to be possible when running in a remote scenario.

**Comment by brunopacheco1** [comment]
I would like to know your opinion on how to tackle this.

Should we revert the changes related to this cob_field evaluation to a previous state where we were able to evaluate alphanumeric and numeric values and start taking care of other usage scenarios, case by case?

**Comment by GitMensch** [comment]
Is https://sourceforge.net/p/open-cobol/feature-requests/383/  and https://sourceforge.net/p/open-cobol/feature-requests/384/ silver bullets? If yes, is there a special way it must be implemented? Note: if it must return a char * then the content must be copied "somewhere" as the only way other than passing a storage in is to use a buffer - and this would be re-used for every request and increase to some pre-defined maximum. So if the debugger only shows what is at the address then it will be already broken with querying the next field...

**Comment by brunopacheco1** [comment]
Definitely having those functions is going to help a lot, as we wouldn't have to deal with any transformation (this extension should be only an interface/broker to the real debugger).

The way it should be implemented, I can't tell much as I am not familiar with C, but if I understood correctly of what you said, by calling `char * cob_field_get_value (cob_field *f)`, I wouldn't display the internal buffer address content, I would display what the function returns when executing a `data-evaluate-expression` command, for instance.

I also understood you liked more this option, `void cob_field_copy_value (char * buffer, size_t size, cob_field *f)`, and it makes sense, as you are delegating to the client the responsibility of handling the buffer. But here I am not sure how I would get the cob_field in one single call of `data-evaluate-expression`.

**Comment by brunopacheco1** [comment]
I'll close this issue in favor to #59 , as that one is specific to the `USAGE` scenario, which is not supported yet by the extension.
